src/MCPClient/lib/clientScripts/pid_declaration.py

```diff
@@ -0,0 +1,208 @@
+                )
+            elif isinstance(mdl, Directory):
+                msg = "Identifier {}: {} added for: {}: {}".format(
+                    identifier_type,
+                    identifier,
+                    "Directory",
+                    os.path.basename(os.path.split(mdl.currentlocation)[0]),
+                )
+            else:
+                msg = "Identifier {}: {} added for: {}: {}".format(
+                    identifier_type,
+                    identifier,
+                    "File",
+                    os.path.basename(mdl.currentlocation),
+                )
+            self.job.pyprint(msg)
+        except KeyError:
+            self.job.pyprint(
+                "Identifier {}:{} added to {}".format(identifier_type, identifier, mdl)
+            )
+
+    def _add_identifier_to_model(self, mdl, id_):
+        """Add custom identifier to the model we found a match for. Adding the
+        identifier may return a AssignPIDsException we simply treat this as a
+        warning and let the script continue. This behavior will be easy to
+        change if required with further usage of this feature.
+        """
+        identifier_type = id_.get("identifierType", None)
+        identifier = id_.get("identifier", None)
+        try:
+            _add_custom_pid_to_mdl_identifiers(mdl, identifier_type, identifier)
+            self._log_identifer_to_stdout(mdl, identifier_type, identifier)
+            self.actual_count += 1
+        except AssignPIDsException as err:
+            self.job.pyprint(err, file=sys.stderr)
+
+    def parse_and_attach_identifiers(self, unit_uuid, json_data):
+        """Take our identifiers and add them to the model objects that we want
+        them to belong to by the time we create the AIP METS.
+        """
+        self.identifier_count = len(json_data)
+        for file_object in json_data:
+            file_path = file_object.get("file")
+            identifiers = file_object.get("identifiers", [])
+            objects_dir, base_name = os.path.split(file_path)
+            if (objects_dir == "objects" or objects_dir == "") and not base_name:
+                file_path = self.SIP_IDENTIFIER
+            else:
+                # Run name normalization to allow us to pair our data with
+                # information already in the database.
+                file_path = self._fixup_path_input_by_user(file_path)
+            for id_ in identifiers:
+                if file_path == self.SIP_IDENTIFIER:
+                    mdl = SIP.objects.get(uuid=unit_uuid)
+                    self._add_identifier_to_model(mdl, id_)
+                    continue
+                try:
+                    mdl = File.objects.get(sip_id=unit_uuid, currentlocation=file_path)
+                    self._add_identifier_to_model(mdl, id_)
+                    continue
+                except File.DoesNotExist:
+                    pass
```

Comment:
Maybe worth logging here?

Reply:
True, if someone has tried to assign a custom identifier to something that doesn't exist then this will be useful. Will put this in, and for the directory too.

Reply:
I didn't want to introduce too much logic/state so as an alternative, I log the things that we can confirm as added, and then provide a count afterwards. I do think we can put a greater emphasis on validation of these files up-front so that users have a better chance of their input matching, but that's a bigger problem with a number of special files like this. Same with directories. 

IDK, is that okay? It will be pretty easy to add if it comes up through user-testing.

Reply:
Yeah, that's totally fine. It's always a trade off between noise and information that's needed in logs.